### PR TITLE
Sample data test infrastructure

### PR DESCRIPTION
**Pull request by j-emberton**
Have added some basic infrastructure to get up and running with domain specific integration level tests:

Added matlab engine pytest fixture that can be configured to run matlab code from the original or refactored code repo.

Added fixture to read canonical test case data into temporary dirs for testing

Added /test_artifacts/ dir within /tests/ to be a home for reference test data.

Added the input and output files associated with the US_ARc_sample_data.

The idea here is that any individual test can demand a fixture which sets up a specific set of reference data for that test 

This should form the backbone of testing against domain data unit and regression tests going forward

### CODE REVIEW

**Comment by j-emberton** [comment]
@dorchard @isaacaka @tztsai 

I've setup the matlab engine envisaging that we have two parallel sets of matlab code - 'ustar_cp' and 'ustar_cp_refactor_wip'. The latter would initially be non-refactored matlab code, but over time would be incrementally refactored as we work on it.

This allows us to work from root to leaf as discussed yesterday with the lower level dependencies remaining intact while we refactor the higher level functions.

I'm hoping with this approach Matlab doesn't get confused with duplicate function names on its path - I haven't tested it yet.

This specific issue wouldn't affect the bulk of the implementation in this PR however.

If you are ok for us to test this approach I'll create the 'ustar_cp_refactor_wip' dir and migrate the matlab code into it.

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import pytest
+import os
+import matlab.engine
+import shutil
+
+@pytest.fixture()
+def matlab_engine(refactored = False):
+    # Start MATLAB engine
```
Comment (dorchard) [inline tests/conftest.py:29]: Add docstring

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    eng = matlab.engine.start_matlab()
+
+    current_dir = os.getcwd()
+    code_path = 'oneflux_steps/ustar_cp/'
```
Comment (dorchard) [inline tests/conftest.py]: Perhaps best to set this to `None` initially to avoid constant duplication with line 16

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    eng.addpath(matlab_function_path, nargout=0)
+
+    yield eng
+
+    # Close MATLAB engine after tests are done
+    eng.quit()
+
+@pytest.fixture
+def setup_folders(tmpdir, testcase):
+    """Fixture to setup input and output folders for tests by copying all files from a specified local directory."""
```
Comment (dorchard) [inline tests/conftest.py]: Add docstring for the inputs

**Review by dorchard — COMMENTED** [review]
A few comments around documentation and a minor tweak but otherwise looks good. I think we probably also need a requirements.txt as well for the tests, or to extend the top-level one, to include `matlabengine` and `shutil`?

File: tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    output_folder = tmpdir.mkdir("output")
+
+    # Path to the local directory containing files to be copied
+    testcase_path = os.join.path('tests/test_fixtures/local_directory', testcase)
+
+    # Ensure the local directory exists
+    if not os.path.exists(testcase_path):
+        raise FileNotFoundError(f"Local directory {testcase_path} does not exist")
+
+    def copy_tree(src, dst):
```
Comment (isaacaka) [inline tests/conftest.py]: Could you use `shutil.copytree()` instead of implementing the function `copy_tree()`?

**Comment by isaacaka** [comment]
Looks good, left a comment on using shutil.copytree()

**Comment by j-emberton** [comment]
> Looks good, left a comment on using shutil.copytree()

Good shout

**Comment by j-emberton** [comment]
> A few comments around documentation and a minor tweak but otherwise looks good. I think we probably also need a requirements.txt as well for the tests, or to extend the top-level one, to include `matlabengine` and `shutil`?

I've extended the top level one for the time being, though matlab engine is only needed transiently. I think shutil is part of python standard lib so no need for an entry in requirements.txt.

**Comment by j-emberton** [comment]
OK, made the changes you guys suggested and created the duplicate matlab dir. Good to go I think.

File: tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
+import pytest
+import os
+import matlab.engine
+import shutil
+
+@pytest.fixture()
```
Comment (casterkay) [inline tests/conftest.py:6]: Perhaps the scope of the fixture needs to be "module".

**Review by casterkay — COMMENTED** [review]
In PR #19 you also created a `conftest.py` file, which one has a newer version?

**Comment by dorchard** [comment]
Can we have a `README.md` in the `tests` directory that explains the requirements and workflow for running the tests? Also which version of Python should these be using wrt. the requirements? @j-emberton

**Comment by j-emberton** [comment]
> Can we have a `README.md` in the `tests` directory that explains the requirements and workflow for running the tests? Also which version of Python should these be using wrt. the requirements? @j-emberton

Yes. But it might fit better in the launch.m PR where I've actually setup the test env and have running tests. Otherwise the readme will refer to stuff that doesn't exist yet.

File: tests/conftest.py (hunk 5 above)
Comment (j-emberton) [reply tests/conftest.py:6]: Not necessary

**Comment by j-emberton** [comment]
> In PR #19 you also created a `conftest.py` file, which one has a newer version?

You're right. This one is the latest one.

**Comment by dorchard** [comment]
(Note should work with Python 3.11)

**Pull request merged**